Add initial gomobile support

divan opened the pull request:
This PR adds initial "drop-in" [gomobile](https://github.com/golang/mobile) support for status-go.
"Drop-in" means it's supposed to be used as a replacement to the current xgo-based `lib` package. I.e. generated `.aar`/`.framework` files should be absolutely the same and require no modification from `status-react` in order to work.

Important changes:
- [x] Add gomobile installation to the `make setup`
- [x] Replace `statusgo-ios` and `statusgo-android` to use gomobile instead of xgo
- [x] Remove `Jenkinsfile-manual`

Closes https://github.com/status-im/status-go/issues/680

divan committed:
Add drop-in lib package for usage with gomobile

divan committed:
Update Makefile to use gomobile

divan committed:
Add README to mobile

divan commented:
Codeclimate doesn't like duplication in `lib` and `mobile` packages (but it's needed for transition). Travis seems to fail due to flack test `TestPeerPoolSimulationSuite/TestMailServerPeersDiscovery`.

adambabik commented on Makefile:
```diff
@@ -171,7 +164,11 @@ xgo:
 	go get github.com/karalabe/xgo
 
-setup: dep-install lint-install mock-install ##@other Prepare project for first build
+gomobile:
```
`gomobile-install`?

divan replied:
make sense, thanks!

adambabik commented on mobile/response.go:
```diff
@@ -0,0 +1,64 @@
+package status
```
We can either move it to `types.go` or change it to `jsonrpc_response.go`. 

It's an issue that we have so many response types but this one was introduced for bindings that are meant to replace JSON-RPC calls. In status-react, commands like `eth_sendTransaction` are intercepted and a binding is called instead. Thanks to this type of response, the response can be easily returned without further processing.

dshulyak commented on mobile/status.go:
```diff
@@ -0,0 +1,338 @@
+		return makeJSONResponse(err)
+	}
+
+	api.RunAsync(func() error { return statusBackend.StartNode(config) })
```
not really related to this PR. but it seems that RunAsync is totally unnecessary, every call is executed in its own thread. so, for example, StartNode can return an actual error, instead of using signals to report that error. and that means - part of the signals are not useful at all. and the other part of signals can be a simple a subscription over RPC.

dshulyak commented on mobile/status.go:
```diff
@@ -0,0 +1,338 @@
+var statusBackend = api.NewStatusBackend()
+
+// All general log messages in this package should be routed through this logger.
+var logger = log.New("package", "status-go/lib")
```
do you want to keep the name of logger (lib part)?

dshulyak commented on mobile/status.go:
```diff
@@ -0,0 +1,338 @@
+	)
+	errString := ""
+
+	defer func() {
```
this defer usage is so confusing. it definitely should be a simple function that accepts error and returns string

dshulyak reviewed (approved):
comments are mostly irrelevant, but what is the thread situation in code generated by gomobile? as I remember we are using a thread pool in objective-c and java, and application seems to depend on it

divan commented:
@dshulyak answering your comments here in bulk: this PR is "drop-in" replacement, I want to change as little as possible in the original implementation. So I guess the followup PR will be a rethinking the interface between status-go and react with gomobile in mind, and most of your comments will be for sure resolved there.

I'm not quite sure I understood the thread pool issue. Can you elaborate on it?

dshulyak commented:
@divan so both versions of the app are running on a thread pool. but recently there were performance issues on android. after investigation it turned out that pool size was too small. i think it was set to the number of cpu cores (or a double of that). raising number of threads helped. i don't know details of those performance problems, but in current implementation there must be a peer pool in andoid/ios bindings. otherwise app won't work. cc @PombeirP in case he wants to provide more details

pedropombeiro commented:
From a quick test, it seems that there are some open questions:
- `make statusgo-ios` needs XCode to run, so we'd need to adapt the Makefiles to run each build on the appropriate host;
- `make statusgo-android` needs some environment variables, so it can find things like the NDK. We need to think what the best way is to go around that. Right now the documentation doesn't reflect that (btw the documentation has moved to docs.status.im in `develop`).

> xgo is not removed yet, because it's probably still needed for status-desktop (to confirm)

If we collectively decide that it is not absolutely required to be able to cross-compile iOS/MacOS libraries, then I'd definitely get rid of xgo, as otherwise I don't see much advantage in bringing gomobile to the equation.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

divan commented:
Thanks for feedback @PombeirP. So let's breakdown the problems we have:
 - Unclear if it's possible to make an iOS gomobile build on Linux machine (Jenkins/CI)
 - Lack of documentation of what to do when Android NDK is not installed 

Is that correct?

Regarding the Android NDK: I had an assumption that `gomobile init` does all the needed checks and validations, so no further env variables needed to be specified, but I have to test this assumption.

pedropombeiro commented:
Yeah, that's correct. Although I'm quite sure that XCode doesn't run on anything but a Mac. When you mentioned the PR was ready for testing on the react side (pending being able to build status-react), I thought you had already tested the two types of builds to confirm that the libraries were generated correctly, but if I understand correctly, that still remains to be done, right?

pedropombeiro commented:
As future context: I'm currently working on [another PR](https://github.com/status-im/status-react/pull/5976) which is meant to provide status-go static library artifacts to Status Desktop builds. I'm running into issues consuming the cross-compiled OSX library, so I'm starting to think that with the move to `gomobile`, we'll already be tied to OSX hosts anyway, so might as well stop relying on `xgo` altogether and ensure that we build OSX artifacts on OSX hosts. cc @jakubgs

divan committed:
Merge branch 'develop' into gomobile

divan committed:
Fix library API

dmitryn commented:
I'm trying to make manual build with `gomobile` branch to integrate it into status-react, but CI fails with:
https://ci.status.im/job/status-go/job/status-go-manual/544/console

```
[status-go] Running shell script
+ make ci
lint
level=warning msg="[runner/megacheck] Can't run megacheck because of compilation errors in packages [github.com/status-im/status-go/mobile]: mobile/status.go:67: LoadNodeConfig not declared by package params and 6 more errors: run `golangci-lint run --no-config --disable-all -E typecheck` to see all errors"
mobile/status.go:67:19: LoadNodeConfig not declared by package params (typecheck)
	_, err := params.LoadNodeConfig(configJSON)
	                 ^
mobile/status.go:43:24: LoadNodeConfig not declared by package params (typecheck)
	config, err := params.LoadNodeConfig(configJSON)
	                      ^
mobile/status.go:28:47: cannot convert "" (untyped string constant) to uint64 (typecheck)
	config, err := params.NewNodeConfig(datadir, "", params.FleetBeta, uint64(networkID))
	                                             ^
Makefile:260: recipe for target 'lint' failed
make: *** [lint] Error 1
[Pipeline] }
[Pipeline] // dir
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 2
Finished: FAILURE
```

@divan is there any instructions on how to build it?

<blockquote><img src="/static/5a99659d/favicon.ico" width="48" align="right"><div><strong><a href="https://ci.status.im/job/status-go/job/status-go-manual/544/console">status-go » status-go-manual #544 Console [Jenkins]</a></strong></div></blockquote>

adambabik commented:
This branch is just outdated as some packages in status-go changed.

divan commented:
I pushed the commit with rebase, please try now.

dmitryn commented:
Failing:

```
[status-go-manual] Running shell script
+ go get github.com/status-im/xgo
[Pipeline] parallel
[Pipeline] [statusgo-android] { (Branch: statusgo-android)
[Pipeline] [statusgo-ios-simulator] { (Branch: statusgo-ios-simulator)
[Pipeline] [statusgo-android] dir
[statusgo-android] Running in /home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go
[Pipeline] [statusgo-android] {
[Pipeline] [statusgo-ios-simulator] dir
[statusgo-ios-simulator] Running in /home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go
[Pipeline] [statusgo-ios-simulator] {
[Pipeline] [statusgo-android] sh
[statusgo-android] [status-go] Running shell script
[statusgo-android] + make statusgo-android
[Pipeline] [statusgo-ios-simulator] sh
[statusgo-android] Installing gomobile...
[statusgo-ios-simulator] [status-go] Running shell script
[statusgo-ios-simulator] + make statusgo-ios-simulator
[statusgo-ios-simulator] make: *** No rule to make target 'statusgo-ios-simulator'.  Stop.
[Pipeline] [statusgo-ios-simulator] }
[Pipeline] [statusgo-ios-simulator] // dir
[Pipeline] [statusgo-ios-simulator] }
[statusgo-ios-simulator] Failed in branch statusgo-ios-simulator
[statusgo-android] Building status-go for Android...
[statusgo-android] gomobile: no Android NDK path is set. Please run gomobile init with the ndk-bundle installed through the Android SDK manager or with the -ndk flag set.
[statusgo-android] Makefile:118: recipe for target 'statusgo-android' failed
[statusgo-android] make: *** [statusgo-android] Error 1
[Pipeline] [statusgo-android] }
[Pipeline] [statusgo-android] // dir
[Pipeline] [statusgo-android] }
[statusgo-android] Failed in branch statusgo-android
[Pipeline] // parallel
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 2
Finished: FAILURE
```

https://ci.status.im/job/status-go/job/status-go-manual/545/console

<blockquote><img src="/static/36a3b6fc/favicon.ico" width="48" align="right"><div><strong><a href="https://ci.status.im/job/status-go/job/status-go-manual/545/console">status-go » status-go-manual #545 Console [Jenkins]</a></strong></div></blockquote>

divan commented:
@dmitryn hmm, so Makefile should have `gomobile init` command to prepare CI environment. Can you manually put this command to pipeline before Makefile is executing?

dmitryn commented:
@divan I've added `gomobile init` before `make statusgo-android` and `make statusgo-ios-simulator`, but it didn't help


```
+ go get github.com/status-im/xgo
[Pipeline] parallel
[Pipeline] [statusgo-android] { (Branch: statusgo-android)
[Pipeline] [statusgo-ios-simulator] { (Branch: statusgo-ios-simulator)
[Pipeline] [statusgo-android] dir
[statusgo-android] Running in /home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go
[Pipeline] [statusgo-android] {
[Pipeline] [statusgo-ios-simulator] dir
[statusgo-ios-simulator] Running in /home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go
[Pipeline] [statusgo-ios-simulator] {
[Pipeline] [statusgo-android] sh
[statusgo-android] [status-go] Running shell script
[statusgo-android] + gomobile init
[Pipeline] [statusgo-ios-simulator] sh
[statusgo-ios-simulator] [status-go] Running shell script
[statusgo-ios-simulator] + sh gomobile init
[statusgo-ios-simulator] sh: 0: Can't open gomobile init
[Pipeline] [statusgo-ios-simulator] }
[Pipeline] [statusgo-ios-simulator] // dir
[Pipeline] [statusgo-ios-simulator] }
[statusgo-ios-simulator] Failed in branch statusgo-ios-simulator
[Pipeline] [statusgo-android] sh
[statusgo-android] [status-go] Running shell script
[statusgo-android] + make statusgo-android
[statusgo-android] Installing gomobile...
[statusgo-android] Building status-go for Android...
[statusgo-android] gomobile: no Android NDK path is set. Please run gomobile init with the ndk-bundle installed through the Android SDK manager or with the -ndk flag set.
[statusgo-android] Makefile:118: recipe for target 'statusgo-android' failed
[statusgo-android] make: *** [statusgo-android] Error 1
[Pipeline] [statusgo-android] }
[Pipeline] [statusgo-android] // dir
[Pipeline] [statusgo-android] }
[statusgo-android] Failed in branch statusgo-android
[Pipeline] // parallel
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 127
Finished: FAILURE
```

divan commented:
@dmitryn that's because `sh gomobile init` failed – `sh` seems to expect sh script as an input, not the command itself.

ghost commented:
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

adambabik committed:
clean up Makefile

adambabik committed:
Merge branch 'develop' into gomobile

adambabik committed:
update missing exports

adambabik committed:
clean up

adambabik committed:
prepare Jenkinsfile

adambabik committed:
fix Jenkinsfile

adambabik committed:
add steps

adambabik committed:
try now

adambabik committed:
fix parentheses

jakubgs committed:
fix location of envrionment section

adambabik committed:
split setup

adambabik committed:
skip TestGetFilterLogs

adambabik committed:
disable USB detection

adambabik committed:
run gomobile-install in setup-build

adambabik committed:
add NoUSB in all places

adambabik committed:
set ANDROID_* envs

adambabik committed:
Merge branch 'develop' into gomobile

adambabik committed:
Merge branch 'develop' into gomobile

adambabik committed:
use ios/arm64 target for ios

mandrigin committed:
rename package from `status` to `statusgo` to generate the iOS framework correctly.

mandrigin commented:
it looks like it generates the correct framework with https://github.com/status-im/status-go/pull/1164/commits/8177afadb1e6d2c2bdb90e0fb0956c7b3e2aca40 now

adambabik commented:
@mandrigin Wohoo! Thanks!

mandrigin committed:
fixup for iOS signals compiling

adambabik committed:
fix ios target

mandrigin committed:
use Go code to propagate signals on mobile

mandrigin committed:
clean up most of the C magic of signals package (only leftovers for Desktop)

mandrigin commented:
@adambabik I removed most of the C dynamic-code-loading magic, and events seems to be able to work on iOS again. I will need to change the Android implementation, will probably do it tomorrow.

adambabik committed:
fix target for android

adambabik committed:
Merge remote-tracking branch 'origin/develop' into gomobile

adambabik committed:
move comment

cammellos committed:
Notify user when the device is missing (#1298)

* Notify user when the device is missing

* Update services/shhext/chat/encryption.go

Co-Authored-By: cammellos <andrea.maria.piana@gmail.com>

mandrigin committed:
make NofifyUsers work again (after rebase)

dshulyak committed:
Mail peer store and connection manager (#1295)

This change implements connection manager that monitors 3 types of events:
1. update of the selected mail servers
2. disconnect from a mail server
3. errors for requesting mail history

When selected mail servers provided we will try to connect with as many as possible, and later disconnect the surplus. For example if we want to connect with one mail server and 3 were selected, we try to connect with all (3), and later disconnect with 2. It will to establish connection with live mail server faster.

If mail server disconnects we will choose any other mail server from the list of selected. Unless we have only one mail server. In such case we don't have any other choice and we will leave things as is.

If request for history was expired we will disconnect such peer and try to find another one. We will follow same rules as described above.

We will have two components that will rely on this logic:
1. requesting history

If target peer is provided we will use that peer, otherwise we will request history from any selected mail server that is connected at the time of request.

2. confirmation from selected mail server

Confirmation from any selected mail server will bee used to send a feedback that envelope was sent.

I will add several extensions, but probably in separate PRs:
1. prioritize connection with mail server that was used before reboot
2. disconnect from mail servers if history request wasn't expired but failed.
3. wait some time in RequestsMessage RPC to establish connection with any mail server

Currently this feature is hidden, as certain changes will be necessary in status-react. 

partially implements: https://github.com/status-im/status-go/issues/1285

mandrigin commented:
@adambabik I believe that the go side is fully functional right now (tested on iOS and Android), so we should be ready to merge it.

adambabik commented:
Follow https://github.com/status-im/status-react/pull/6914#issuecomment-444583772.

adambabik commented:
@mandrigin I think we need to coordinate with https://github.com/status-im/status-react/pull/6914. Otherwise, it won't be possible to compile and use status-go HEAD in other status-react builds.

mandrigin commented:
@adambabik I'm not sure the current bundling scripts work nowadays, because the paths have been changed, but yeah.

adambabik committed:
delete Jenkinsfile-manual

ghost commented:
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

mandrigin commented:
Nope, this one is still to be done.

ghost commented:
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

mandrigin commented:
still active

adambabik committed:
Merge branch 'develop' into gomobile

adambabik committed:
add fixes after rebase

adambabik committed:
fix Desktop stage in Jenkinsfile

adambabik committed:
fix Desktop archive in Jenkinsfile

adambabik committed:
fix status.go

adambabik committed:
Merge branch 'develop' into gomobile

ghost commented:
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

gravityblast commented:
I think it's still active

adambabik committed:
Merge branch 'develop' into gomobile

adambabik committed:
fix mobile bindings

adambabik commented:
There is a problem with compilation for Android using gomobile. The issue we experience was reported here https://github.com/golang/go/issues/29706. It needs to be fixed in Android NDK and the issue is https://github.com/android-ndk/ndk/issues/849. 

EDIT: it builds fine on iOS: https://ci.status.im/blue/organizations/jenkins/status-go%2Fparallel/detail/parallel/263/pipeline/

mandrigin commented:
Weird, it build okay on my computer.... Let's see...

mandrigin commented:
it is for sure builds just fine with NDK r16b, let's see what we have on our Jenkins runners...

jakubgs commented:
@mandrigin 
We use r10e: https://github.com/status-im/status-react/blob/develop/.TOOLVERSIONS#L1

<blockquote><img src="https://avatars3.githubusercontent.com/u/11767950?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react">status-im/status-react</a></strong></div><div>a free (libre) open source, mobile OS for Ethereum - status-im/status-react</div></blockquote>

mandrigin commented:
@jakubgs it doesn't even pass `gomobile init` if I give it r10e on my computer... I'll try to build it in Docker, let's see how it goes....

jakubgs commented:
@mandrigin then maybe it's about time we upgraded the NDK...

mandrigin commented:
@jakubgs okay, the simplest and safest way to proceed is:
- to install `r16b` to a separate folder (let's name it `/opt/android-ndk-r16b`) to the builder machines;
- to export `/opt/android-ndk-r16b` as an ENV variable (`$NDK_GOMOBILE`);
- add parameter to `gomobile init` -> `gomobile init -ndk $NDK_GOMOBILE`;
- bingo! 🎉

jakubgs commented:
>to install r16b to a separate folder (let's name it /opt/android-ndk-r16b) to the builder machines;

That's not how we do things anymore. It's all part of a docker container.

mandrigin commented:
@jakubgs that include mac machines, I don't see how we can build gomobile for iOS in Docker...

mandrigin commented:
@jakubgs anyway, how hard is it to make the NDK r16b available separately and have an environment variable? Then I can just update the setup script and it will work.

jakubgs commented:
Okay, You can access the NDK at `/opt/android-ndk-r16b` on all MacOS hosts.
Change: https://github.com/status-im/infra-misc/commit/f0e5f9f71c737dea111f05658749d803c302ff81

jakubgs commented:
And also added the `NDK_GOMOBILE` env variable for `jenkins` user:
Change: https://github.com/status-im/infra-misc/commit/bae916c99921cbe392533770515bd38b55e50929

mandrigin committed:
Use custom NDK for building gomobile.

mandrigin committed:
(to be reverted) disable canary test

mandrigin committed:
(to be reverted) only build mobile release on MacOS

mandrigin committed:
Revert "(to be reverted) only build mobile release on MacOS"

This reverts commit cd5f884e56f1af01d6bbc8f63c02c7380d8fc70c.

mandrigin committed:
fix packages folder conflict in Jenkinsfile

mandrigin committed:
one more attempted fixup for Jenkinsfile

mandrigin committed:
update version (`g` means `gomobile`, update major because backward compat)

jakubgs committed:
use stash/unstash to release at the end

Signed-off-by: Jakub Sokołowski <jakub@status.im>

mandrigin committed:
Fix how we stash Statusgo.framework

Signed-off-by: Jakub Sokołowski <jakub@status.im>

mandrigin committed:
Tweak "Release" phase.

mandrigin committed:
fix syntax error in Jenkinsfile (eeh)

mandrigin committed:
one more syntax fixup

mandrigin committed:
:(

mandrigin committed:
add debug output

mandrigin committed:
add cleanup code

mandrigin committed:
fix dir path

mandrigin committed:
make Android stash look more like iOS

mandrigin committed:
Clean-up after yourself.

mandrigin committed:
more debug output

mandrigin committed:
even more debug info

mandrigin committed:
add more debug info

mandrigin committed:
try not to clean-up files after stashing

mandrigin committed:
Merge branch 'develop' into gomobile

mandrigin committed:
Add `UpdateMailservers` to gomobile exports.

jakubgs committed:
split builds into separate Jenkinsfiles

Signed-off-by: Jakub Sokołowski <jakub@status.im>

mandrigin committed:
Merge remote-tracking branch 'origin/gomobile-split' into gomobile

mandrigin committed:
bump version to .1

mandrigin commented:
android: works

mandrigin commented:
iOS: works.

mandrigin commented:
thanks @divan for starting it!
@adambabik for wrapping it up!
@jakubgs for fighting dear god scary nightmares of Jenkins!
